test/Npgsql.Tests/DataSourceTests.cs

```diff
@@ -380,4 +381,86 @@ public async Task ConfigureJsonOptions_is_order_independent()
+
+        Assert.Throws<InvalidCastException>(() => connection.ExecuteScalar($"SELECT 'happy'::{type}"));
+
+        // Close connection and reopen to make sure it picks up the new type and mapping from the data source
+        connection.Close();
+        connection.Open();
+
+        Assert.DoesNotThrow(() => connection.ExecuteScalar($"SELECT 'happy'::{type}"));
+    }
+
+    [Test]
+    public async Task ReloadTypesAsync()
+    {
+        await using var adminConnection = await OpenConnectionAsync();
+        var type = await GetTempTypeName(adminConnection);
+
+        await using var dataSource = CreateDataSourceWithType(type);
+        await using var connection = await dataSource.OpenConnectionAsync();
+
+        await connection.ExecuteNonQueryAsync($"CREATE TYPE {type} AS ENUM ('sad', 'ok', 'happy')");
+        await dataSource.ReloadTypesAsync();
+
+        Assert.ThrowsAsync<InvalidCastException>(async () => await connection.ExecuteScalarAsync($"SELECT 'happy'::{type}"));
+
+        // Close connection and reopen to make sure it picks up the new type and mapping from the data source
+        await connection.CloseAsync();
+        await connection.OpenAsync();
+
+        Assert.DoesNotThrowAsync(async () => await connection.ExecuteScalarAsync($"SELECT 'happy'::{type}"));
+    }
+
+    [Test]
+    public async Task ReloadTypesAsync_across_data_sources()
+    {
+        await using var adminConnection = await OpenConnectionAsync();
+        var type = await GetTempTypeName(adminConnection);
+
+        await using var dataSource1 = CreateDataSourceWithType(type);
+        await using var connection1 = await dataSource1.OpenConnectionAsync();
+
+        await using var dataSource2 = CreateDataSourceWithType(type);
+        await using var connection2 = await dataSource2.OpenConnectionAsync();
+
+        await connection1.ExecuteNonQueryAsync($"CREATE TYPE {type} AS ENUM ('sad', 'ok', 'happy')");
+        await dataSource1.ReloadTypesAsync();
+
+        Assert.ThrowsAsync<InvalidCastException>(async () => await connection1.ExecuteScalarAsync($"SELECT 'happy'::{type}"));
+        Assert.ThrowsAsync<InvalidCastException>(async () => await connection2.ExecuteScalarAsync($"SELECT 'happy'::{type}"));
+
+        // Close connection and reopen to check that the new type and mapping is not available in dataSource2
+        await connection2.CloseAsync();
+        await connection2.OpenAsync();
+
+        Assert.ThrowsAsync<InvalidCastException>(async () => await connection2.ExecuteScalarAsync($"SELECT 'happy'::{type}"));
+
+        await dataSource2.ReloadTypesAsync();
+
+        // Close connection2 and reopen to make sure it picks up the new type and mapping from dataSource2
+        await connection2.CloseAsync();
+        await connection2.OpenAsync();
+
+        Assert.DoesNotThrowAsync(async () => await connection2.ExecuteScalarAsync($"SELECT 'happy'::{type}"));
+    }
+
+    private NpgsqlDataSource CreateDataSourceWithType(string type)
```

Comment:
I'd personally just inline this, doesn't seem worth it for 3 lines

Reply:
OK